### ISSUE TITLE
build rstudio successfully but cannot find qt libs

### ISSUE THREAD

**Issue by beew** [rstudio/rstudio#3061]
RStudio Edition : Desktop
RStudio Version : 1.2.xx (release tar ball) and have tried master, same result
OS Version  : Ubuntu 18.04 64 bit
R Version :  3.50 (and 3.44 before,same result)

Compile rstudio from source, run ./install-dependencies-debian script to download all the dependencies including qt-sdk which is version 5.10 (only glitch was some error about not able to cd into plumer or something like that, but just run ./install-qt-sdk maually and it worked)

rstudio builds successfully but it won't start. when run in terminal, get

```
/usr/local/lib/rstudio/bin/rstudio: /usr/lib/x86_64-linux-gnu/libQt5Core.so.5: version `Qt_5.10' not found (required by /usr/local/lib/rstudio/bin/rstudio)
```
exporting LD_LIBRARY=/path/to/rstudio's qt-dk:$LD_LIBRARY_PATH ( in my system it is /home/myusername/Qt5.10.1/5.10.1/gcc_64/lib) get different error

```
This application failed to start because it could not find or load the Qt platform plugin "xcb"
in "".

Reinstalling the application may fix this problem.
Aborted (core dumped)
```

**Comment by kevinushey** [comment]
How did you build and install RStudio? Note in particular that we have a [script here](https://github.com/rstudio/rstudio/blob/master/package/linux/make-package) that we use for building package builds on Linux.

Of particular note is the various CMake definitions that get set during configure time here:

https://github.com/rstudio/rstudio/blob/master/package/linux/make-package#L63-L75

Are you setting these all in your own local build as well?

**Comment by beew** [comment]
I built it according to the instructions in the INSTALL file
1) Run install-dependencies-debian script
2) create build directory in the root rstudio directory
3) run cmake with target Desktop, install prefix /usr/local
4) make
5) make install (actually I used checkinstall which created and install a .deb but that should have no bearing on it

I have been doing this for a long time on Ubuntu 16.04 without problem (but for rstudio 1.1x branch which won't build in 18.04 since missing libgstreamer0.10), the question is which qt it actually links to, it seems that even in 16.04 it is linked to system's qt rather than downloaded qt-sdk since I always delete it after building rstudio with no consequence. on my Ubuntu 16.04 machine
```
$ldd /usr/local/lib/rstudio/bin/rstudio --verbose
/usr/local/lib/rstudio/bin/rstudio:
	linux-vdso.so.1 =>  (0x00007ffe3c75a000)
	/usr/lib/liblapack.so (0x00007fbcb98a8000)
	libQt5WebKitWidgets.so.5 => /usr/lib/x86_64-linux-gnu/libQt5WebKitWidgets.so.5 (0x00007fbcbbfac000)
	libQt5PrintSupport.so.5 => /usr/lib/x86_64-linux-gnu/libQt5PrintSupport.so.5 (0x00007fbcbbf39000)
	libQt5DBus.so.5 => /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5 (0x00007fbcbbeba000)
	libQt5WebKit.so.5 => /usr/lib/x86_64-linux-gnu/libQt5WebKit.so.5 (0x00007fbcb745e000)
	libQt5Network.so.5 => /usr/lib/x86_64-linux-gnu/libQt5Network.so.5 (0x00007fbcb7304000)
	libQt5Widgets.so.5 => /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5 (0x00007fbcb6c77000)
	libQt5Gui.so.5 => /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5 (0x00007fbcb672f000)
	libQt5Core.so.5 => /usr/lib/x86_64-linux-gnu/libQt5Core.so.5 (0x00007fbcb6259000)
```

**Comment by kevinushey** [comment]
Normally RStudio should be linking to bundled versions of those libraries, e.g.

```
$ ldd /usr/lib/rstudio/bin/rstudio | grep libQt5
	libQt5Core.so.5 => /usr/lib/rstudio/bin/../lib/libQt5Core.so.5 (0x00007f0478d72000)
	libQt5Widgets.so.5 => /usr/lib/rstudio/bin/../lib/libQt5Widgets.so.5 (0x00007f04784cb000)
	libQt5Gui.so.5 => /usr/lib/rstudio/bin/../lib/libQt5Gui.so.5 (0x00007f0477c8c000)
	libQt5Network.so.5 => /usr/lib/rstudio/bin/../lib/libQt5Network.so.5 (0x00007f04778e1000)
	libQt5WebChannel.so.5 => /usr/lib/rstudio/bin/../lib/libQt5WebChannel.so.5 (0x00007f04776bf000)
	libQt5WebEngineWidgets.so.5 => /usr/lib/rstudio/bin/../lib/libQt5WebEngineWidgets.so.5 (0x00007f0477472000)
	libQt5PrintSupport.so.5 => /usr/lib/rstudio/bin/../lib/libQt5PrintSupport.so.5 (0x00007f04771f7000)
	libQt5DBus.so.5 => /usr/lib/rstudio/bin/../lib/libQt5DBus.so.5 (0x00007f0476f65000)
	libQt5WebEngineCore.so.5 => /usr/lib/rstudio/bin/../lib/libQt5WebEngineCore.so.5 (0x00007f047087e000)
	libQt5Qml.so.5 => /usr/lib/rstudio/bin/../lib/../lib/libQt5Qml.so.5 (0x00007f046be34000)
	libQt5Quick.so.5 => /usr/lib/rstudio/bin/../lib/../lib/libQt5Quick.so.5 (0x00007f046b788000)
	libQt5Positioning.so.5 => /usr/lib/rstudio/bin/../lib/../lib/libQt5Positioning.so.5 (0x00007f046b4ec000)
	libQt5QuickWidgets.so.5 => /usr/lib/rstudio/bin/../lib/../lib/libQt5QuickWidgets.so.5 (0x00007f046b2d6000)
```

Did those supporting Qt libraries get properly installed on your system? (Note that you need to set `-DRSTUDIO_PACKAGE_BUILD=1` to ensure that RStudio bundles + uses its own version of Qt)

**Comment by beew** [comment]
> Note that you need to set -DRSTUDIO_PACKAGE_BUILD=1 to ensure that RStudio bundles + uses its own version of Qt

Ah, I didn't do that, will try again and get back.

However, I did try
```
cmake .. -DRSTUDIO_TARGET=Desktop -DCMAKE_BUILD_TYPE=Release -DQT_QMAKE_EXECUTABLE=$HOME//Qt5.10.1/5.10.1/gcc_64/bin/qmake
```
with no luck

**Comment by beew** [comment]
This time I did
```
cmake .. -DRSTUDIO_TARGET=Desktop -DCMAKE_BUILD_TYPE=Release -DRSTUDIO_PACKAGE_BUILD=1
...
**-- Found Qt: /home/bee/Qt5.10.1/5.10.1/gcc_64/bin/qmake**
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bee/Downloads/rstudio-1.2.679/build
``
So it found the bundled QT
but after install, rstudio still doesn't start complaining that 
```
/usr/local/lib/rstudio/bin/rstudio: /usr/lib/x86_64-linux-gnu/libQt5Core.so.5: version `Qt_5.10' not found (required by /usr/local/lib/rstudio/bin/rstudio)
```
ldd /usr/local/lib/rstudio/bin/rstudio --verbose
/usr/local/lib/rstudio/bin/rstudio:
/usr/local/lib/rstudio/bin/rstudio: /usr/lib/x86_64-linux-gnu/libQt5Core.so.5: version `Qt_5.10' not found (required by /usr/local/lib/rstudio/bin/rstudio)
```
But this time with
```
 export LD_LIBRARY_PATH=/home/bee/Qt5.10.1/5.10.1/gcc_64/lib:$LD_LIBRARY_PATH
```
It does work, so the runtime library still needed to be specified even if rstudio is built against its bundled version of Qt, is this expected?

Also I don't know why the cmake flag

```-DQT_QMAKE_EXECUTABLE=$HOME/Qt5.10.1/5.10.1/gcc_64/bin/qmake
```
didn't work before, since it seems to be the same as what 
```-DRSTUDIO_PACKAGE_BUILD=1
```
from the cmake output above.

**Comment by kevinushey** [comment]
No, that's not expected -- a package build should bundle and use the Qt libraries installed alongside RStudio.

Here's where that bundling occurs in the CMake sources:

https://github.com/rstudio/rstudio/blob/bb450e15a604530f1f442b88d1cb366dfab7cc80/src/cpp/desktop/CMakeLists.txt#L524-L531

And here's what the associated script does (it runs when the Debian package is built)

https://github.com/rstudio/rstudio/blob/bb450e15a604530f1f442b88d1cb366dfab7cc80/src/cpp/desktop/install/LinuxInstallQt.sh.in

Effectively, it makes sure that all of the requisite Qt libraries are bundled with RStudio, and also does some hackery with the RPATH to ensure RStudio always looks for Qt in the right place.

**Comment by beew** [comment]
> No, that's not expected -- a package build should bundle and use the Qt libraries installed alongside RStudio.

So it is a bug then?

**Comment by kevinushey** [comment]
Maybe a bug in the documentation. Ultimately, the primary mechanism for building a redistributable package for RStudio is through [make-package](https://github.com/rstudio/rstudio/blob/master/package/linux/make-package) so if anything this would just be a bug in the documentation.

**Comment by gtritchie** [comment]
I did (on Ubuntu 16):
- `cmake .. -DRSTUDIO_PACKAGE_BUILD=1 -DRSTUDIO_TARGET=Desktop -DCMAKE_BUILD_TYPE=Release`
- `sudo make install`

and compared to an installation via DEB. All the files were there, in the correct locations (including the qt.conf files), but the problem remains that the `make install` instance doesn't find it's Qt.

We may ultimately just update the INSTALL documentation to recommend building and installing via `make-package` (and also mention that we provide Dockerfiles to make it easier to do so). Still, if it's a simple tweak to fix `make install` it would be _nice_.

**Comment by datamole** [comment]
@gtritchie I had posted the related bug report #3128 that you just joined.

Thanks for your tip to compile the package using the make-package script. That worked on my opensuse tumbleweed system. However compilation into a custom directory using cmake should ideally also get fixed for Linux dostros without supported package system.

**Comment by gtritchie** [comment]
Yes, we switched our dependency script to use the online Qt installer. It is considerably larger, but the variations between our manually tweaked bundled packages and the way we setup our dev machines were troublesome, so we opted to unify on a single way to install.

I don't believe our scripts have ever detected a system Qt install? I could be mistaken there, though. If you run the dependency script with --exclude-qt-sdk it skips the Qt install altogether.
